cmd/vm/switch_linux.go

```diff
@@ -0,0 +1,230 @@
+	// network namespace (rd1), the logic behind this approach was the
+	// lack of support for network namespaces in the AF_VSOCK libs.
+	connFile := os.NewFile(uintptr(3), "RancherDesktop-AFVsock-Connection")
+	defer connFile.Close()
+
+	logrus.Debugf("using a AF_VSOCK connection file from default namespace: %v", connFile)
+
+	// this should never happen
+	if err := checkForExsitingIf(tapIface); err != nil {
+		logrus.Fatal(err)
+	}
+
+	tap, err := water.New(water.Config{
+		DeviceType: water.TAP,
+		PlatformSpecificParams: water.PlatformSpecificParams{
+			Name: tapIface,
+		},
+	})
+	if err != nil {
+		logrus.Fatalf("creating tap device %v failed: %s", tapIface, err)
+	}
+	logrus.Debugf("created tap device %s: %v", tapIface, tap)
+	defer func() {
+		tap.Close()
+		logrus.Debugf("closed tap device: %s", tapIface)
+	}()
+
+	if err := linkUp(tapIface, defaultMacAddr); err != nil {
+		logrus.Fatalf("setting mac address [%s] for %s tap device failed: %s", defaultMacAddr, tapIface, err)
+	}
+	if err := loopbackUp(); err != nil {
+		logrus.Fatalf("enabling loop back device failed: %s", err)
+	}
+
+	logrus.Debugf("setup complete for tap interface %s(%s) + loopback", tapIface, defaultMacAddr)
+
+	for {
+		if err := run(connFile, tap); err != nil {
+			logrus.Error(err)
+		}
+		time.Sleep(time.Second)
+	}
+}
+
+func run(connFile *os.File, tap *water.Interface) error {
+	errCh := make(chan error, 1)
+	go tx(connFile, tap, errCh, defaultMTU)
+	go rx(connFile, tap, errCh, defaultMTU)
+	go func() {
+		if err := dhcp(tapIface); err != nil {
+			errCh <- fmt.Errorf("dhcp error: %w", err)
+		}
+	}()
+
+	return <-errCh
+}
+
+func loopbackUp() error {
+	lo, err := netlink.LinkByName("lo")
+	if err != nil {
+		return err
+	}
+
+	return netlink.LinkSetUp(lo)
+}
+
+func linkUp(iface, mac string) error {
+	link, err := netlink.LinkByName(iface)
+	if err != nil {
+		return err
+	}
+	if mac == "" {
+		return netlink.LinkSetUp(link)
+	}
+	hw, err := net.ParseMAC(mac)
+	if err != nil {
+		return err
+	}
+	if err := netlink.LinkSetHardwareAddr(link, hw); err != nil {
+		return err
+	}
+
+	logrus.Debugf("successful link setup %+v\n", link)
+	return netlink.LinkSetUp(link)
+}
+
+func dhcp(iface string) error {
+	if _, err := exec.LookPath("udhcpc"); err == nil { // busybox dhcp client
+		cmd := exec.Command("udhcpc", "-f", "-q", "-i", iface, "-v")
+		cmd.Stderr = os.Stderr
+		cmd.Stdout = os.Stdout
+		return cmd.Run()
+	}
+	cmd := exec.Command("dhclient", "-4", "-d", "-v", iface)
+	cmd.Stderr = os.Stderr
+	cmd.Stdout = os.Stdout
+	return cmd.Run()
+}
+
+func rx(conn io.Writer, tap *water.Interface, errCh chan error, mtu int) {
+	logrus.Info("waiting for packets...")
+	var frame ethernet.Frame
+	for {
+		frame.Resize(mtu)
+		n, err := tap.Read([]byte(frame))
+		if err != nil {
+			errCh <- fmt.Errorf("reading packet from tap failed: %w", err)
+			return
+		}
+		frame = frame[:n]
+
+		size := make([]byte, 2)
+		binary.LittleEndian.PutUint16(size, uint16(n))
```

Comment:
Hmm, is the frame in little-endian? I thought most network things were big-endian (but I haven't checked, but if it's working it's _probably_ correct?)